test/AutoDiff/Sema/derivative_attr_type_checking.swift

```diff
@@ -1165,3 +1150,41 @@ func opaqueResult(_ x: Float) -> some Differentiable { x }
 func vjpOpaqueResult(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
   fatalError()
 }
+
+// Test instance vs static method mismatch.
+
+struct StaticMismatch<T: Differentiable> {
+  // expected-note @+1 {{original function 'init(_:)' operates on a type}}
+  init(_ x: T) {}
+  // expected-note @+1 {{original function 'instanceMethod' operates on an instance type}}
+  func instanceMethod(_ x: T) -> T { x }
+  // expected-note @+1 {{original function 'staticMethod' operates on a type}}
+  static func staticMethod(_ x: T) -> T { x }
+
+  @derivative(of: init)
```

Comment:
In these diagnostics we should make the error appear at the location that triggered the error. For example, in this case `vjpInit(_:)` is a perfectly fine Swift function if it were compiled directly. The compilation error is caused by the addition of the `@derivative(of: init)` attribute, so the error should appear on the attribute, not on `vjpInit(_:)`.
```console
@derivative(of: init)
                ^~~~
```
The error should also say "as required by _what_" instead of just "as required" in order to provide the user with the sufficient context. In this case, maybe it can say something like:
```console
error: unexpected derivative function declaration; 'StaticMismatch.init(_:)' requires that the derivative function be 'static'
```

Reply:
yes, you are right. It seems a bit weird to have a note produced "far away" from the error itself. I'll address your comment in a soon-to-come commit :)

Reply:
I have modified the wording of the error messages and notes to be more aligned to what you suggested. The error diagnostic is now pointing at the method stated in the 'of:' parameter.

Reply:
> ```
> @derivative(of: init)
>                 ^~~~
> ```

I overlooked your suggestion here of highlighting the original function (sorry for that :/ ). I've seen other diagnostics in this context do highlighting as well so I just pushed #36259 for the sake of consistency.